### ISSUE TITLE
MW Logic got busted somewhere...

### ISSUE THREAD

**Issue by LostThePirate** [OoTRandomizer/OoT-Randomizer#581]
Dev v4.2.6, P2's first logical Bow in P1's Ganon's Castle. All Medallions seed, P2's Forest is a Medallion.
Previously, MW logic would prevent this situation, by not placing items in places that are inaccessible to both the sender and receiver of items distributed between worlds. Something between 4.0 and 4.2.6 broke this function.
(Also, apparently I can no longer drop a Spoiler Log directly here, as "We don't support that file type" is what pops when I attempt to drop this .json in. It can be found here in the main OoTR Discord: https://discordapp.com/channels/274180765816848384/438698093354156032/551700192311640064 )

**Comment by r0bd0g** [comment]
Rename the tile to .txt

**Comment by r0bd0g** [comment]
Can we put in a request to somehow make these spoilers reachable by a human? It's really hard to parse at the moment.

**Comment by BeeLuigi** [comment]
This might not be something that was introduced post-4.0. Stuff like this can happen with the current mw balance fix implementation.

Let's say that GC bow was the first P2 bow being placed. It doesn't see any problem with that, because both players can reach that location (because P2 still has two unplaced bows), so it places it there.
From then on that bow can be collected by the maximum_exploration_state_list with just the access requirements of P1. Which means that as long as P1 can reach the GC bow, P2's other two bows can be placed in locations that are locked behind it, making this one WotH. 
Not sure if that's what happened here, but that shouldn't be too hard to figure out.

**Comment by BeeLuigi** [comment]
[OoT_F5B64_04H9KI6WY7_W2_Spoiler.txt](https://github.com/TestRunnerSRL/OoT-Randomizer/files/2923137/OoT_F5B64_04H9KI6WY7_W2_Spoiler.txt)

Example of it happening in a pre 4.0 seed even. P1's Boomerang is at P2's Ganon's Castle Fairy. The boomerang eventually locks a P1 strength upgrade in GTG.

**Comment by r0bd0g** [comment]
The playthrough and hints are not beholden to the multi rebalance. I'm not going to hyper-analyze that seed but it may be that another route exists to acquire the strength upgrade.

**Comment by TestRunnerSRL** [comment]
For Lost, looking through the spoiler, it looks like the bow that was in the "balanced" logic was from OoT, and the items required for player 2 are available. So this seems fine. As r0bd0g said, the playthrough listed is just one of many possible playthroughs, and does not try to make it is the playthrough that always plays in the balance.

As far as BeeLuigi, the balance algorithm fails when dealing with item chains. They are generally pretty unlikely, but can happen. I'm not sure the best way to handle that issue, and is known.

**Comment by r0bd0g** [comment]
"For Lost, looking through the spoiler, it looks like the bow that was in the "balanced" logic was from OoT"
That Bow from player 2 in Ganon's Castle is Way of the Hero, so either the items to get the Bow from OoT are not truly available, or the WotH is messed up. Either way there's a bug here.

That Bow for player 2 I'm seeing in this spoiler is at Mask of Truth, not on OoT, which is a disabled location. I guess another possibility is that there's some weird interaction with multi-balance and disabled locations.

**Comment by Zannick** [comment]
Lost's spoiler has P2's bows at: P1 GC, P4 GC Light Trial, and P3 Mask of Truth (disabled). P4 has a strength upgrade at a P4 SoS Grotto. P4's SoS is in Sheik at Kakariko, and Bongo Bongo has their Forest Medallion, but P4's Hover Boots are behind P2's 10 Big Poes. So P4 GC Light Trial requires the P1 GC P2 Bow, and disabled locations aren't considered in can_beat_game, so the P1 GC location is correctly WotH.

Bee's item chain is P2 GC Fairy (P1 Boomerang), P1 Barinade (P2 Hover Boots), P2 Shadow Temple Boss Key (P1 Hookshot), P1 Lake Hylia Sun (P2 Gerudo Card), P2 GTG (P1 Strength).

The maximum_exploration_state_list is generated by giving each player all unplaced items and then running collect_locations, which will resolve item chains (even absurd ones like this), and then check reachability only of whatever it's placing. So I could imagine if it placed the P1 Strength in that chain last: generates maximum_exploration_state_list including unplaced P1 Gerudo Card and P2 Strength 3, then the chain resolves and P2 has Gerudo Card as well, so both P1 and P2 can reach the GTG chest. If it placed the Gerudo Card last, then the P2 GTG isn't reachable in maximum_exploration_state_list, but all that's asked is: can both P1 and P2 reach Lake Hylia Sun? That inaccessible P1 Strength isn't necessary for P2 to get the P1 Hookshot in that chain, so yes. And so on.

Seems to me like the problem is that it's hard to verify that the receiving player can reach a given location at the same time as the sending player, when the reachability for the receiving player can be changed by items placed later.

**Comment by r0bd0g** [comment]
That's absurdly janky. Can you imagine trying to keep track of the item placement logic in a multiworld like this... yikes...

**Comment by TestRunnerSRL** [comment]
@Zannick Well it's not so much that the conditions are changing, but it's that initially there are items that are assumed reachable and later are no assumptions. So depending on the item placement order, that assumptions might resolve to True or False. For normal item placement, we make sure to place items such that the the assumptions will always resolve True, but this isn't maintained for the balance. TBH as I've said before, I don't have a problem with the items occasionally not satisfying the balance. The purpose of the balance isn't so that you can make deductions about the item placement, it's more to reduce the amount of wait-time between players. So I've never really considered this a bug nor have I wanted to try to fix this, unless you have any suggestions.

@r0bd0g I have no idea what you're talking about.

**Comment by r0bd0g** [comment]
Me neither I guess.

**Comment by Zannick** [comment]
Yes, I was just trying to clarify the effect and the desired behavior. The desired behavior as described in the code comments makes it sound like you did want the balance to hold across all item placements. The effect of checking the balance at distribution time only is that it's more likely to hold for items placed later (things accessible early). Anyway, I don't see an easy way to verify balance will hold when all the items are placed except by verifying after all the items are placed. (Like, check that a player has locations in every playthrough sphere until they win?) Or rewriting the distribution algorithm, definitely not worth the effort.
